WIP: Propose centralized experiment tracking in Kubeflow (#1)
mprahl/kubeflow-community: proposals/NNNN-experiment-tracking/README.md

# KEP-NNNN: Centralize Experiment Tracking in Kubeflow

## Summary

This proposal aims to resolve the current fragmented and limited experiment tracking experience by expanding the
**Kubeflow Model Registry** into a unified, centralized metadata store. Currently, experiment tracking is scattered
across components like Kubeflow Pipelines (which requires pipeline execution for tracking) and Katib (limited to
hyperparameter tuning). This leads to challenges such as **limited flexibility** for direct logging from Python scripts
or Jupyter notebooks, a **fragmented user experience** across multiple interfaces, and **maintenance difficulties** due
to reliance on the inactive MLMD (Machine Learning Metadata) project.

The proposal tackles these issues by:

- **Expanding Model Registry** into a central experiment tracking store for experiments, runs, metrics, and artifacts
  across all Kubeflow components.

> - Providing **MLFlow SDK compatibility**, enabling users to leverage familiar APIs while storing data in Kubeflow.

Reviewer: Maybe rephrase this along the lines of "supporting third party" experimentation tracking. Mention MLFlow as a possible target instead of a goal?

maintaining consistent metadata throughout the ML lifecycle. This has the added benefit of eliminating technical debt
associated with Kubeflow Pipelines' heavy reliance on MLMD.

## Motivation

Currently, experiment tracking in Kubeflow is fragmented and tightly coupled to specific components. Kubeflow Pipelines
requires users to run everything through pipelines to track experiments, while Katib's experiments are limited to
hyperparameter tuning scenarios. This fragmentation creates several challenges:

1. **Limited flexibility**: Users cannot log experiments directly from Python scripts, Jupyter Notebooks, or other
   Kubeflow components without forcing everything through pipelines. This restriction often drives users to seek
   solutions outside the Kubeflow ecosystem.
1. **Fragmented experience**: Users must navigate multiple interfaces to correlate run results and evaluation data,
   depending on which Kubeflow component they use (e.g., Pipeline runs, Katib experiments, Training Operator results).
1. **No unified tracking**: Users cannot easily compare runs, share insights, or maintain consistent metadata across the
   entire Kubeflow ecosystem.

> 1. **Maintenance challenges**: The Kubeflow Pipelines dependency on MLMD (Machine Learning Metadata) creates technical

Reviewer 2: ```suggestion
1. **Maintenance challenges**: The Kubeflow Pipelines dependency on MLMD creates technical
```

   debt and limits future flexibility due to its lack of active maintenance. MLMD also overlaps with the Model Registry
   objectives defined in the
   [Model Registry overview](https://www.kubeflow.org/docs/components/model-registry/overview/#use-case-2-experimenting-with-different-model-weights-to-optimize-model-accuracy),
   creating separate metadata store experiences.

These limitations make it difficult to effectively manage the machine learning lifecycle across Kubeflow.

### Goals

> 1. **Kubeflow centralized experiment tracking**: Create an experiment tracking system that is independent from Kubeflow

Reviewer: maybe merge first and second goals, and rephrase to emphasize separation of concerns and experiment tracking as a first class citizen capability in Kubeflow?

   Pipelines, enabling experiments to be logged and managed without requiring pipeline execution.
1. **Transform Model Registry into a unified metadata store**: Evolve the existing Model Registry component into a
   centralized metadata store that can handle experiments, runs, metrics, and artifacts across the entire Kubeflow
   ecosystem.

> 1. **Pipelines integration**: Integrate Kubeflow Pipelines with the centralized experiment tracking system in Model

Reviewer: rephrase emphasizing refactoring pipelines to externalize experiment tracking, removing the dependency (tech debt) from Pipelines. This allows pipelines to integrate with MR as well as 3rd party trackers.

   Registry, allowing pipeline runs and output artifacts to be automatically logged and tracked.

> 1. **Remove the MLMD dependency from Pipelines**: Decouple Kubeflow Pipelines from MLMD (Machine Learning Metadata) to

Reviewer 2: ```suggestion
1. **Remove the MLMD dependency from Pipelines**: Decouple Kubeflow Pipelines from MLMD to
```

   remove a core dependency on a project outside of the Kubeflow ecosystem.
1. **Extend SDK capabilities for experiment tracking**: Enhance the Kubeflow SDK to support direct logging of
   experiments, runs, metrics, and artifacts without requiring pipeline execution.
1. **Provide MLFlow SDK compatibility**: Create an

>    [MLFlow SDK plugin](https://mlflow.org/docs/latest/ml/plugins/#storage-plugins) that allows users to leverage the

Reviewer 3: It should be `MLflow plugin` instead, since these extend MLflow's backend storage capabilities rather than the SDK itself
```suggestion
   [MLFlow plugin](https://mlflow.org/docs/latest/ml/plugins/#storage-plugins) that allows users to leverage the
```

>    [MLFlow SDK plugin](https://mlflow.org/docs/latest/ml/plugins/#storage-plugins) that allows users to leverage the

Reviewer: I am of two minds about explicitly setting this as a goal, since the plugin is standalone. Having the MLFlow plugin as a separate enhancement might make it easier to approve, both.

   familiar MLFlow SDK experiment tracking APIs while working with the centralized Kubeflow experiment tracking
   solution.
1. **Implement multitenancy in Model Registry**: Add multitenancy support to Model Registry to enable isolation at the
   Kubernetes namespace level like Kubeflow Pipelines does today. This resolves the multitenancy gap that exists with
   MLMD today.
1. **Differentiate Katib experiments in the Kubeflow dashboard**: Update the Kubeflow dashboard to clearly distinguish
   between Katib's hyperparameter tuning experiments and general Kubeflow experiments, improving user experience and
   reducing confusion.

### Non-Goals

1. **Decoupling Katib's experiments from their current implementation**: While the Kubeflow community can revisit
   integration with Katib after the initial implementation, we are not targeting the decoupling of Katib's existing
   experiment tracking system as part of this proposal. The initial focus is on the SDK and Pipelines experience.
1. **MLFlow UI support**: Although the MLFlow SDK plugin should theoretically enable MLFlow UI compatibility, providing

>    direct MLFlow UI support is not a targeted goal of this proposal.

Reviewer 3: Why theoretically? The plugin should definitely enable MLflow UI compatibility with Model Registry -- that's the purpose of MLflow plugins. You just need to install the MLflow package and point the UI to the Model Registry backend (with the Model Registry plugin installed), same as you would with other data sources like local files, databases, etc.
```suggestion
1. **MLFlow UI support**: Although the MLFlow plugin should theoretically enable MLFlow UI compatibility, providing
   direct MLFlow UI support is not a targeted goal of this proposal.
```

Reviewer: I think splitting the MLFlow plugin into a separate proposal will help as it removes this kind of issue from the critical path

Author: @kramaranya it was mostly due to some of the comments Dhiraj made about some UI issues he encountered when implementing the plugin but I'll remove the word theoretical.

Author: @etirelli I see your point but I like keeping it together to provide the reader the whole vision and to also ensure the new domain models in the Model Registry map well to MLFlow ones to enable this compatibility. Implementation of the MLFlow plugin can be separate and done later depending on the design/implementation of the Kubeflow SDK working group.

> 1. **Automatic experiment tracking from the Kubeflow Training operator**: While users can leverage the new functionality

Reviewer 2: ```suggestion
1. **Automatic experiment tracking from the Kubeflow Trainer**: While users can leverage the new functionality
```

   through the Kubeflow SDK or MLFlow SDK plugin, we are not implementing automatic experiment tracking directly from

>    the Kubeflow Training operator as part of this proposal.

Reviewer 2: ```suggestion
   the Kubeflow Trainer as part of this proposal. The community should revisit integrations at a later date.
```

>    the Kubeflow Training operator as part of this proposal.

Reviewer 3: To clarify: The MLflow plugin allows saving metadata to the Model Registry backend when using MLflow APIs.

"Kubeflow SDK or MLflow SDK plugin" is confusing, since we could either use Kubeflow SDK or MLflow SDK , with MLflow plugin regardless of which SDK you choose. However, I would just keep Kubeflow SDK here to not confuse users.
```suggestion
1. **Automatic experiment tracking from the Kubeflow Training operator**: While users can leverage the new functionality
   through the Kubeflow SDK with MLflow plugin, we are not implementing automatic experiment tracking directly from
   the Kubeflow Training operator as part of this proposal.
```

## Proposal

### Model Registry Expansion

Model Registry's current focus is managing metadata for the lifecycle of machine learning models, from registration and
versioning through deployment and serving. The proposal is to expand the scope to cover experiment tracking to be a

> centralized metadata store for Kubeflow. As part of this expansion of scope, we should consider renaming it to reflect

Reviewer 2: ```suggestion
centralized metadata store for Kubeflow. As part of this expansion of scope, we could consider renaming it to reflect
```

- **Artifact Management**: Metadata storage of models, datasets, and other ML artifacts
- **Unified Metadata**: Consistent metadata schema across the entire Kubeflow ecosystem

### Kubeflow Pipelines Integration

The integration of Kubeflow Pipelines with the centralized experiment tracking system will be implemented in two phases:

#### Phase 1: Dual Registration and Unified Experiment View

- Enable Kubeflow Pipelines to register experiment metadata in both MLMD and Model Registry simultaneously
- Provide users with a unified view of experiments across the entire Kubeflow ecosystem
- Maintain backward compatibility with existing MLMD-dependent tooling and existing Kubeflow Pipelines UI

#### Phase 2: Complete MLMD Decoupling

- Eliminate the MLMD dependency by migrating all Kubeflow Pipelines (KFP) metadata storage to the KFP database and
  external metadata trackers
- Provide configuration options allowing users to choose between Model Registry and MLMD for experiment tracking
- Kubeflow Pipelines UI enhancements with more direct integration with Model Registry if that mode is chosen

### SDK Integration

#### MLFlow SDK Compatibility

> - Implement MLFlow SDK plugins to connect to the Model Registry backend

Reviewer 3: ```suggestion
- Implement MLFlow plugins to connect to the Model Registry backend
```

Author: Thanks! I'll remove SDK from these occurrences.

- Enable users to leverage MLFlow's familiar APIs while storing data in centralized Kubeflow infrastructure
- Support automatic logging for popular ML frameworks (TensorFlow, PyTorch, scikit-learn, etc.) through MLFlow's SDK

#### Kubeflow SDK Enhancement

- Provide a more native, Kubeflow-centric experience
- Simplify setup and configuration

> - Integrate with existing Model Registry SDK functionality

Reviewer 3: How does this sound?
```suggestion
#### Kubeflow SDK Enhancement

- Provide native experiment tracking capabilities within the Kubeflow SDK using MLflow compatibility
- Simplify experiment tracking setup and configuration
- Enable seamless integration with Model Registry for unified metadata management
- Support both local and remote experiment tracking workflows
```

Reviewer: +1 to the suggestions, except, remove the MLFlow reference

Author: Great suggestion. Thank you @kramaranya!

### Experiment Tracking UI

The UI will provide a comprehensive interface for experiment management and visualization with feature parity to the
existing Kubeflow Pipelines run comparison functionality, while extending support to all run sources generically.

Key UI components will include:

- Experiments overview page with filtering and sorting capabilities
- Experiment detail view with hierarchical run structures
- Run management interface with advanced filtering options
- Compare runs with multiple chart types

### User Stories

#### Story 1: Data Scientist Logging Experiments from Jupyter Notebooks

As a data scientist working with Kubeflow, I want to log my experiments directly from Jupyter Notebooks without being
forced to use pipelines, so that I can track my iterative model development process naturally. I should be able to use

> familiar MLFlow SDK APIs to log metrics, parameters, and artifacts, and then view all my experiments in a unified

Reviewer: will we not have kubeflow SDK support? are we talking only about MLFlow?

Author: Good point. I'll split this into two user stories.

Kubeflow dashboard alongside pipeline runs and other ML workflows.

#### Story 2: ML Engineer Comparing Runs Across Different Sources

As an ML engineer, I want to compare experiment runs that come from different sources (Jupyter notebooks, pipeline runs,
training jobs) in a single unified interface, so that I can make informed decisions about model performance regardless
of how the experiments were executed. I should be able to filter runs by experiment, parameters, metrics, and source,
then visualize comparisons using parallel coordinates plots, scatter plots, and other chart types.

#### Story 3: Platform Administrator Managing Multi-tenant Experiment Tracking

As a platform administrator, I want to ensure that experiment tracking data is properly isolated by Kubernetes namespace
using RBAC, so that different teams can securely share the same Kubeflow cluster without data leakage. I should be able
to configure namespace-level access controls for experiments, runs, and artifacts, while maintaining the ability to
provide global visibility for cross-team collaboration when needed.

### Risks and Mitigations

> 1. **Migration Challenges**: The proposal involves migrating away from MLMD (Machine Learning Metadata), which is

Reviewer 2: ```suggestion
1. **Migration Challenges**: The proposal involves migrating away from MLMD, which is
```

   currently used by Kubeflow Pipelines. This migration process could be complex and risky, potentially breaking
   existing workflows. To mitigate this risk, we will implement extensive testing and maintain an extended transition
   period where Kubeflow Pipelines registers to both MLMD and Model Registry.
1. **Dependency on External Plugin**: The proposal relies heavily on MLFlow SDK compatibility through plugins, creating
   a dependency on MLFlow's plugin architecture. This dependency could become fragile if MLFlow changes their plugin
   APIs or if the plugins don't work as expected. To mitigate this risk, the Kubeflow SDK will also support these APIs,
   allowing us to move away from wrapping MLFlow without breaking user code. Furthermore, popular plugins such as from
   [Microsoft Azure](http://docs.azure.cn/en-us/machine-learning/how-to-use-mlflow-configure-tracking) would cause
   significant disruption for the MLFlow community if they were to change their SDK plugin stance, making such changes
   unlikely.

> ## Design Details

Reviewer 3: Shall we add a digram to visualize the proposal design, similar to what we had for f2f. What do you think?

Author: Good suggestion! I'll create one before submitting this to the Kubeflow community.

### Model Registry Domain Models

The expanded Model Registry will include the following domain models, heavily influenced by
[kubeflow/model-registry#1224](https://github.com/kubeflow/model-registry/issues/1224#issuecomment-3068005968). This is
meant to be a high-level design with further refinement being done in a Model Registry specific KEP or design document.

> 1. **Experiments**: A group of related runs. This would replace experiments in Pipelines and could map to experiments in

Reviewer 2: ```suggestion
1. **Experiments**: A group of related runs. This would replace experiments in Pipelines and could potentially map to experiments in
```

   tasks.
1. **Images**: An artifact used to represent an image to be rendered directly in the browser. This will be mainly used
   for custom visualizations not available natively in the UI.

### Filtering Requirements

Some basic level of filtering is required for initial UI rendering. This includes but is not limited to:

- Filter all entities by namespace
- Filter runs by experiment name
- Filter runs by parameter name
- Filter runs by metric name
- Filter runs by parent run
- Filter runs by status
- Filter runs by metadata fields/tags
- Filter runs by source
- Filter all entities by user that created it
- Filter by timestamps (before, after)
- Fuzzy searching on entity names

### Multi-tenancy Implementation

Model Registry's current authorization model gates access at the API level through a proxy rather than at the individual
resource or namespace level. This approach prevents the same Model Registry instance from being shared across teams that
require isolation. This limitation conflicts with Kubeflow Pipelines' multiuser mode, which is the default deployment

> strategy for the Kubeflow platform. MLMD (ML Metadata) has the same limitation, so adding multi-tenancy support to Model

Reviewer 2: ```suggestion
strategy for the Kubeflow platform. MLMD has the same limitation, so adding multi-tenancy support to Model
```

- Gradual migration path that doesn't require immediate MLMD deprecation
- Global experiment visibility across all Kubeflow components
- Maintains backward compatibility with existing MLMD-dependent tooling

#### Phase 2: Complete MLMD Decoupling

**Objective**: Makes MLMD optional by migrating all Kubeflow Pipelines metadata storage to either the KFP database and
Model Registry/MLMD (user chooses), providing users with flexibility in their experiment tracking solution.

**Implementation Details**:

- Migrate all MLMD-stored state beyond experiment tracking to the Kubeflow Pipelines database
- Provide configuration options allowing users to choose between Model Registry and MLMD as their experiment tracking
  backend
- Modify the KFP UI to be more integrated with Model Registry

**Benefits**:

- Eliminates a hard dependency on the not actively maintained MLMD project
- Provides users with choice between Model Registry and MLMD based on their specific needs
- Reduces technical debt and maintenance burden
- Aligns with the broader Kubeflow ecosystem's metadata strategy

### SDK Implementation Details

> #### MLFlow SDK Model Registry Plugin

Reviewer 3: ```suggestion
#### MLFlow Model Registry Plugin
```

1. **Tracking Store Plugin**: This plugin will handle experiment and run metadata, metrics, parameters, and tags. It
   will map MLFlow's tracking concepts to the Model Registry's domain models (Experiments, Runs, Users, Metrics, etc.)
   while excluding tracing functionality for the initial implementation.

2. **Artifact Repository Plugin**: This plugin will manage artifact storage and retrieval, leveraging the Model
   Registry's artifact management capabilities to store model files, datasets, and other ML artifacts.

With these plugins in place, users will be able to use the MLFlow SDK with minimal changes to their existing code. They
simply need to install an additional Python package and specify the Model Registry API as their tracking URI. This
approach enables users to leverage MLFlow's powerful
[automatic logging](https://mlflow.org/docs/latest/ml/tracking/autolog/) functionality for popular ML frameworks
(TensorFlow, PyTorch, scikit-learn, etc.) while storing all data in the centralized Kubeflow Model Registry.

**Future Enhancements**: In subsequent phases, we may extend the plugin architecture to include:

> - **Tracing Support**: Adding tracing capabilities would require new domain models in Model Registry.

Reviewer 2: ```suggestion
- **Tracing Support**: Adding tracing capabilities akin to [MLflow](https://mlflow.org/docs/latest/genai/tracing/)would require new domain models in Model Registry.
```

- **Model Registry Store**: Further development could expose the Model Registry model registration functionality in the
  MLFlow SDK.

The primary objective remains enabling users to leverage MLFlow's mature ecosystem and familiar APIs while benefiting
from Kubeflow's centralized experiment tracking infrastructure.

#### Kubeflow SDK Implementation

> Building on the MLFlow SDK Model Registry plugin, we can enhance the Kubeflow SDK to provide a more native,

Reviewer 2: ```suggestion
Building on the MLFlow SDK Model Registry plugin, we can enhance the Kubeflow SDK to provide a native
```

> Kubeflow-centric experience. The initial implementation would focus on simplifying setup by automatically configuring

Reviewer 3: ```suggestion
Building on the MLFlow Model Registry plugin, we can enhance the Kubeflow SDK to provide a native,
Kubeflow-centric experience. The initial implementation would focus on simplifying setup by automatically configuring
```

> the MLFlow SDK Model Registry plugin and creating convenient wrapper functions around the supported MLFlow SDK APIs.

Reviewer 3: ```suggestion
the MLFlow Model Registry plugin and creating convenient wrapper functions around the supported MLFlow SDK APIs.
```

This integration should feel cohesive with the existing Model Registry SDK functionality for registering models.

> Enhancements could include:
> 
> - **Authentication**: Provide helpers for handling authentication with the Model Registry service, such as generating
>   Kubernetes tokens.
> - **Integrated UI Access**: Provide direct links to the Model Registry UI from within the Kubeflow SDK, making it easier
>   for users to visualize and manage their experiments.
> - **Discovery Services**: Automatically detect available Model Registry installations on the cluster and allow users to
>   select one.

Reviewer 3: ```suggestion
Enhancements could include:

- **Authentication**: Provide helpers for handling authentication with the Model Registry service, such as generating
  Kubernetes tokens.
- **Integrated UI Access**: Provide direct links to the Model Registry UI from within the Kubeflow SDK, making it easier
  for users to visualize and manage their experiments.
- **Discovery Services**: Automatically detect available Model Registry installations on the cluster and allow users to
  select one.
- **Local Experiments**: Enable users to work with local experiments and selectively publish them to the remote Model Registry backend.
```

Author: Good suggestion! Thank you.

# KEP-NNNN: Centralize Experiment Tracking in Kubeflow

## Summary

This proposal aims to resolve the current fragmented and limited experiment tracking experience by expanding the
**Kubeflow Model Registry** into a unified, centralized metadata store. Currently, experiment tracking is scattered
across components like Kubeflow Pipelines (which requires pipeline execution for tracking) and Katib (limited to
hyperparameter tuning). This leads to challenges such as **limited flexibility** for direct logging from Python scripts
or Jupyter notebooks, a **fragmented user experience** across multiple interfaces, and **maintenance difficulties** due
to reliance on the inactive MLMD project.

The proposal tackles these issues by:

> - **Expanding Model Registry** into a central experiment tracking store for experiments, runs, metrics, and artifacts

Reviewer 4: Model registry was designed from the ground up to be a metadata store for AI/ML platforms. So adding support for experiments is just a matter of adding new metadata resource types to the existing API. 
The work for this has already been completed to a large extent in model registry component in the following issues and related PRs:
[Add support for Experiment tracking in Model Registry](https://github.com/kubeflow/model-registry/issues/1224)
[Add MLflow SDK support for Model Registry as a store](https://github.com/kubeflow/model-registry/issues/1225)

Author: Thanks! I'll add a section in `### Model Registry Expansion` to mention the original intent of how Model Registry was designed.

   ecosystem, enabling experiments to be logged and managed without requiring pipeline execution.
1. **Pipelines refactoring for externalized experiment tracking**: Refactor Kubeflow Pipelines to externalize experiment
   tracking functionality, removing the built-in MLMD dependency (technical debt) and enabling integration with both
   Model Registry, MLMD, and third-party experiment tracking systems. This decoupling allows pipelines to focus on
   orchestration while experiment tracking becomes a pluggable component.
1. **Extend SDK capabilities for experiment tracking**: Enhance the Kubeflow SDK to support direct logging of
   experiments, runs, metrics, and artifacts without requiring pipeline execution.
1. **MLFlow SDK Integration**: Develop an
   [MLFlow storage plugin](https://mlflow.org/docs/latest/ml/plugins/#storage-plugins) that allows users to continue
   using familiar MLFlow APIs while storing experiment data in Kubeflow's centralized tracking system. This provides
   seamless integration for existing MLFlow code and positions Kubeflow's SDK as complementary to MLFlow rather than

>    competitive.

Reviewer 4: Model Registry MLFlow SDK integration will start with Tracking Store plugin initially to support MLflow experiments API.
https://github.com/kubeflow/model-registry/issues/1225
The intent is to allow existing MLflow users the ability to integrate with Kubeflow platform registry. It is not meant to replace the existing Kubeflow Model Registry SDK which will have full support for Kubeflow model registry Experiments, Registry, and Model Catalog APIs.

Author: I'll reword it to mention the Model Registry SDK and to specify the tracking store plugin explicitly.

1. **Implement multitenancy in Model Registry**: Add multitenancy support to Model Registry to enable isolation at the
   Kubernetes namespace level like Kubeflow Pipelines does today. This resolves the multitenancy gap that exists with

>    MLMD today.

Reviewer 4: The current Model Registry deployment architecture supports multiple model registry instances, which can be shared across teams working in multiple namespaces. Tenancy and isolation is achieved through standard Kubernetes RBAC rules to control access to model registry REST API from different namesapces. 
But, a model registry instance itself is not aware of or enforce namespace level isolation. This makes access control and authorization an orthogonal and therefore flexible concern. 
In other words, Model Registry supports multiple tenants by using an instance per tenant. Applications in a tenant may be distributed across namespaces. A model registry instance or service does not differentiate clients by namespace. 
If a pipeline wishes to record the namespace where it was executed, it will need to add that to either an existing metadata property or use a custom metadata property.

Author: I think this is related to the discussion in https://github.com/mprahl/kubeflow-community/pull/1#issuecomment-3129258661. So let's consolidate that discussion to that thread.

   a dependency on MLFlow's plugin architecture. This dependency could become fragile if MLFlow changes their plugin
   APIs or if the plugins don't work as expected. To mitigate this risk, the Kubeflow SDK will also support these APIs,
   allowing us to move away from wrapping MLFlow without breaking user code. Furthermore, popular plugins such as from
   [Microsoft Azure](http://docs.azure.cn/en-us/machine-learning/how-to-use-mlflow-configure-tracking) would cause
   significant disruption for the MLFlow community if they were to change their SDK plugin stance, making such changes
   unlikely.

## Design Details

### Model Registry Domain Models

The expanded Model Registry will include the following domain models, heavily influenced by
[kubeflow/model-registry#1224](https://github.com/kubeflow/model-registry/issues/1224#issuecomment-3068005968). This is
meant to be a high-level design with further refinement being done in a Model Registry specific KEP or design document.

1. **Experiments**: A group of related runs. This would replace experiments in Pipelines and could potentially map to
   experiments in Katib.

> 1. **Users**: A user to associate runs, metrics, artifacts, and etc. for auditing. This will generally map to the

Reviewer 2: ```suggestion
1. **Users**: A user to associate runs, metrics, artifacts, etc. for auditing. This will generally map to the
```

   files, logs, CSV files, and other data files.
1. **Unregistered Models**: Model Registry currently has the concept of registered model versions, but there is an
   intermediate state that needs to be represented where a run produces an output model that isn't yet registered. It
   has a name, URI, and arbitrary metadata/tags as key-value pairs.
1. **Datasets**: An artifact used for representing input data for training, evaluation, RAG, and other machine learning
   tasks.
1. **Images**: An artifact used to represent an image to be rendered directly in the browser. This will be mainly used
   for custom visualizations not available natively in the UI.

### Filtering Requirements

Some basic level of filtering is required for initial UI rendering. This includes but is not limited to:

- Filter all entities by namespace
- Filter runs by experiment name
- Filter runs by parameter name
- Filter runs by metric name
- Filter runs by parent run
- Filter runs by status
- Filter runs by metadata fields/tags
- Filter runs by source
- Filter all entities by user that created it
- Filter by timestamps (before, after)
- Fuzzy searching on entity names

> ### Multi-tenancy Implementation

Reviewer 4: Model Registry already support multiple tenants with dedicated instance per tenant deployment model. Tenants can span multiple namespaces and access control is enforced through Kubernetes RBAC for the Model Registry service.

Author: I personally consider that a workaround rather than supporting multiple tenants and it would be a significant user experience and maintenance burden to require a new Model Registry instance for every isolated Kubeflow namespace. I'd like to try to work together towards a solution on this in this thread: https://github.com/mprahl/kubeflow-community/pull/1#issuecomment-3129258661

Model Registry's current authorization model gates access at the API level through a proxy rather than at the individual
resource or namespace level. This approach prevents the same Model Registry instance from being shared across teams that
require isolation. This limitation conflicts with Kubeflow Pipelines' multiuser mode, which is the default deployment
strategy for the Kubeflow platform. MLMD has the same limitation, so adding multi-tenancy support to Model Registry
would address this gap for all of Kubeflow's metadata needs.

This proposal adds an authorization mode to Model Registry that leverages Kubernetes subject access review. In this

> mode, all entities would be mapped to Kubernetes namespaces on the cluster to provide namespace-level isolation. For

Reviewer 4: We cannot go back to the namespace scoped limitation of mlmd architecture before. A registered model is not tied to a specific namespace, but shared across all clients within a higher level tenant scope. Model Registry will implement multi-tenancy but use Kubernets user groups and roles to do so, which decouples client location (namespace) from resource location.

Author: The proposal is to have an authorization mode that is namespaced. So in that mode, a namespace would be required on all entities. When not in that mode, it's not a field that is exposed/accepted by the API.

We could start out by having this mode only apply to the new experiment tracking entities but still have the existing entities be global to the instance.

- **Tracing Support**: Adding tracing capabilities akin to [MLflow](https://mlflow.org/docs/latest/genai/tracing/) would
  require new domain models in Model Registry.
- **Model Registry Store**: Further development could expose the Model Registry model registration functionality in the
  MLFlow SDK.

The primary objective remains enabling users to leverage MLFlow's mature ecosystem and familiar APIs while benefiting
from Kubeflow's centralized experiment tracking infrastructure.

#### Kubeflow SDK Implementation

Building on the MLFlow Model Registry plugin, we can enhance the Kubeflow SDK to provide a native Kubeflow-centric
experience. The initial implementation would focus on simplifying setup by automatically configuring the MLFlow Model
Registry plugin and creating convenient wrapper functions around the supported MLFlow SDK APIs. This integration should
feel cohesive with the existing Model Registry SDK functionality for registering models.

Enhancements could include:

- **Authentication**: Provide helpers for handling authentication with the Model Registry service, such as generating
  Kubernetes tokens.

> - **Integrated UI Access**: Provide direct links to the Model Registry UI from within the Kubeflow SDK, making it easier

Reviewer 5: This seems to suggest that the experiments UI will be part of the model Registry UI - is that correct? Or is the intention for the Experiments UI to be a standalone component?

Author: Yes, expanding the MR UI to include experiments was the intent. Thanks for pointing out that this wasn't clear. My latest push adds this clarification.

(TensorFlow, PyTorch, scikit-learn, etc.) while storing all data in the centralized Kubeflow Model Registry.

**Future Enhancements**: In subsequent phases, we may extend the plugin architecture to include:

- **Tracing Support**: Adding tracing capabilities akin to [MLflow](https://mlflow.org/docs/latest/genai/tracing/) would
  require new domain models in Model Registry.
- **Model Registry Store**: Further development could expose the Model Registry model registration functionality in the
  MLFlow SDK.

The primary objective remains enabling users to leverage MLFlow's mature ecosystem and familiar APIs while benefiting
from Kubeflow's centralized experiment tracking infrastructure.

#### Kubeflow SDK Implementation

Building on the MLFlow Model Registry plugin, we can enhance the Kubeflow SDK to provide a native Kubeflow-centric
experience. The initial implementation would focus on simplifying setup by automatically configuring the MLFlow Model
Registry plugin and creating convenient wrapper functions around the supported MLFlow SDK APIs. This integration should
feel cohesive with the existing Model Registry SDK functionality for registering models.

> Enhancements could include:

Reviewer 5: I'm trying to understand the distinction between what the Kubeflow SDK should provide versus what the model registry SDK will provide. My understand is that the MR SDK will contain the mlflow plugin, plus an updated client sdk to reflect the new REST APIs. The 4 points below for me would more naturally live in the model registry SDK as they are all Registry/Experiment specific features.
I think specifying the scope of features for the Model Registry SDK will allow us to know what to build upon in the Kubeflow SDK. I think this will avoid duplication of effort.

Reviewer 6: I echo @briangallagher 's comment and I would like to avoid duplication of efforts in the Kubeflow SDK.

I understand you want to use the MLFlow plugin for experiment tracking, and that is covered by the great work done by @dhirajsb and @syntaxsdev for the mlflow plugin.

When it comes to RegisteredModel, ModelVersion, ModelArtifact etc and [lifecycle of Model Registry user operations](https://model-registry.readthedocs.io/en/latest/#basic-usage), we have already the [MR py sdk](https://pypi.org/project/model-registry/), and this can be integrated in the Kubeflow SDK scope.

Reviewer 4: > I'm trying to understand the distinction between what the Kubeflow SDK should provide versus what the model registry SDK will provide. My understand is that the MR SDK will contain the mlflow plugin, plus an updated client sdk to reflect the new REST APIs. 

The MLFlow plugin won't be part of the MR SDK per se, it's just a module users can install to integrate with MR using MLflow SDK. 

The MR SDK will have it's own APIs for Experimentation as well as the existing RegisteredModel resources. So, I agree that duplicating Experimentation support in a separate Kubeflow SDK doesn't add value. Kubeflow SDK could instead focus on making it easier to locate MR services and configure an MR SDK client that users could use  to access MR API.

Author: @dhirajsb when you say "The MR SDK will have it's own APIs for Experimentation", are you saying the OpenAPI  generated client code or custom methods like `log_metric`?

If it's the OpenAPI generated code, my thought is that the Kubeflow SDK can be the integration point for Model Registry and the Model Registry MLFlow plugin like the POC. We later create bespoke methods like `log_metric` in the Model Registry SDK and replace those transparently in the Kubeflow SDK if we need more functionality.
